README
Hosts3D 1.15 README - 10 May 11  Del Castle
-------------------------------------------
Website
=======
http://hosts3d.sourceforge.net/

Introduction
============
Hosts3D is a 3D real-time network monitor, displaying hosts and packet traffic. Features include support for multiple sensors, analysis of packets to gather hostnames and services, configurable layout of subnetworks, recording/replaying of packet traffic, and the ability to filter packets by hosts, protocol or port.

hsen (Hosts3D Sensor) is a packet capture agent which reads and sends packet header information to Hosts3D, locally or remotely. hsen also equates hostname to IP by reading DNS packets (UDP type A class IN standard query response). Multiple sensors can send information to multiple computers running Hosts3D on the same subnet via broadcast.

License
=======
GNU General Public License Version 2

Hardware Requirements
=====================
Scroll Mouse
Video Card supporting OpenGL (with drivers installed)

Software Requirements
=====================
Linux: libglfw and libpcap (and their associated dependencies)
  Compile: g++, libglfw-dev and libpcap-dev (and their associated dependencies)

Mac OS X: GLFW (http://www.glfw.org/)
  Compile: Mac OS X SDK with GCC and GLFW

FreeBSD: Standard install (hsen only)

Windows: WinPcap (http://www.winpcap.org/)
  Compile: MinGW with GLFW and WinPcap Developer's Pack

Installation
============
Extract tarball: tar xzvf hosts3d-1.15.tar.gz
                 cd hosts3d-1.15

Linux: ./configure
       make
       make install (as root, sudo)
Uninstall: make uninstall (as root, sudo)
Alternative: ./compile-hsen (to create executable "hsen")
             ./compile-hosts3d (to create executable "hosts3d")
             Put executables "hsen" and "hosts3d" in /usr/local/bin/
             Put man pages "hsen.1" and "hosts3d.1" in /usr/local/share/man/man1/

Mac OS X: ./compile-hsen (to create executable "hsen")
          ./compile-mac-hosts3d (to create executable "hosts3d")
          Put executables "hsen" and "hosts3d" in /usr/local/bin/
          Put man pages "hsen.1" and "hosts3d.1" in /usr/local/share/man/man1/

FreeBSD: ./compile-hsen (to create executable "hsen")
         Put executable "hsen" in /usr/bin/
         Put man page "hsen.1" in /usr/share/man/man1/

Windows: Run "compile-hsen.bat" to create executable "hsen.exe"
         Run "compile-hosts3d.bat" to create executable "Hosts3D.exe"

Firewall Configuration: hsen talks to Hosts3D via UDP port 10111.

Starting
========
Starting order does not matter, however if hsen is running and Hosts3D is not, ICMP Port Unreachable (UDP port 10111) may be generated.

Run Hosts3D: hosts3d [-f]
             -f - Display full screen

Run hsen: hsen <id> <interface/file> [<destination>] [-p] [-d]
          id - Identify packets from a specific hsen when multiple exist (1-255)
          interface - Listen on interface (Linux: eth0, eth1, ppp0, wlan0, etc.); or
          file - Read packets from pcap file. Standard input is used if file is "-"
          destination - Hosts3D IP or broadcast address (default localhost)
          -p - Enable promiscuous mode (default disable)
          -d - Run as daemon (default disable)

Windows: hsen -d
         -d - Display interfaces

         hsen <id> <interface/file> [<destination>] [-p]

To capture packets from some interfaces, or in promiscuous mode, you may have to start hsen as root (sudo). In Windows Vista, run as Administrator, even to display interfaces.

For Linux/Mac OS X/FreeBSD, Hosts3D and hsen both log to syslog, check when troubleshooting.

Data Files
==========
In Linux/Mac OS X, created in directory ".hosts3d". In Windows, created in directory "hsd-data".

Hosts3D: controls.txt - Controls
         settings-hsd - Settings
         0network.hnl - Network Layout On-Exit
         1network.hnl - Network Layout 1
         2network.hnl - Network Layout 2
         3network.hnl - Network Layout 3
         4network.hnl - Network Layout 4
         netpos.txt - CIDR Notation Net Position/Colour
         traffic.hpt - Packet Traffic Record
         tmp-hinfo-hsd - Temporary Information
         tmp-netpos-hsd - Temporary Net Positions
         tmp-flist-hsd - Temporary Working Directory File List

Net Positions
=============
If a host is not a member of any net position entries, it is placed in the Grey Zone. If a host is a member of multiple net position entries, the first entry is used. Line format for net position entries is "pos net x-position y-position z-position colour", eg. "pos 123.123.123.123/32 10 0 -10 green".

Positions: Grey/Red - positive x-position
           Blue/Green - negative x-position
           Up - positive y-position
           Down - negative y-position
           Grey/Blue - positive z-position
           Red/Green - negative z-position

Colours: none (where multiple colours are used), default (grey), orange, yellow, fluro, green, mint, aqua, blue, purple, violet and hold (place hosts in same position).

Start/Stop Local hsen in Hosts3D (Linux/Mac OS X)
=================================================
If sudo is required to start/stop a local hsen, the user starting Hosts3D must be in /etc/sudoers. The default command to stop a local hsen is "killall hsen", which will kill all hsen processes.

Hosts3D Help
============
Press H key in Hosts3D to show help.

Notes
=====
- Support only for IPv4.
- IP headers with options are ignored.
- Support for packets with optionless GRE or VLAN 802.1Q encapsulation.
- IP protocol 249 unassigned, used to identify ARP packet.
- IP protocol 250 unassigned, used to identify fragmented IP packet.
- In Hosts3D, by default hosts are added from packet source IP address, activate Add Destination Hosts to also add from destination IP address.
- In Hosts3D, anomalies are a new host or host service.
- In Hosts3D, using the menu options to move/arrange a few thousand hosts may take a few minutes.
- In Hosts3D, deleting hosts, using the menu options to move/arrange hosts, or clearing/restoring a net layout may affect packet recording.
- In Hosts3D, clicking a multiple host object will cycle through selecting hosts within.
- In Hosts3D, data files from 32-bit version are incompatible with 64-bit version and vice versa.
- In Windows, Hosts3D will stall when running a system command on selection, until commands complete. Command output is displayed in associated Command Prompt window.

Reporting Bugs
==============
Report bugs by adding an issue on here, and I'll see if I can fix it.

Copyright
=========
Copyright (c) 2006-2011  Del Castle


ChangeLog
1.15 (10 May 11)
 - (All)     Fixed MinGW static linking in Windows so DLLs aren't required
 - (All)     Removed file INSTALL
 - (Hosts3D) Changed to static linking for GLFW so DLL isn't required
 - (Hosts3D) Removed compiler option -lXrandr (which I thought was needed for GLFW 2.7)
1.14 (13 Apr 11)
 - (All)     Fixed compiler errors with latest MinGW in Windows
 - (All)     Fixed Linux 'make' script showing library options twice
 - (Hosts3D) Compiles with GLFW 2.7
 - (Hosts3D) Added 2D GUI scrollbars
 - (Hosts3D) Fixed show IP/name text not disappearing
 - (Hosts3D) Disable 'show simulated broadcasts' if 'add destination hosts' is enabled, and vice versa
 - (Hosts3D) Added move item up and down buttons to edit net positions
 - (hsen)    Fixed compiler errors with FreeBSD 8.2
 - (hsen)    By default don't run as a daemon, option -d to enable
1.13 (26 Jul 10)
 - (Hosts3D) Stop simulating broadcasts for 255.255.255.255
 - (Hosts3D) Fixed broadcast mask always 0
 - (Hosts3D) Place cursor for text input at left-click
1.12 (12 Jul 10)
 - (Hosts3D) Shift key works with non-alphabetic keys
 - (Hosts3D) Y key now activates automatic link lines for all hosts
 - (Hosts3D) Ctrl + Y key toggles automatic link lines for new hosts, L on OSD
 - (Hosts3D) J key activates automatic link lines for selection
 - (Hosts3D) Ctrl + J key stops automatic link lines for selection
 - (Hosts3D) Ctrl + R key deletes link lines for all hosts
 - (Hosts3D) P on OSD for show packets for new hosts (instead of N)
 - (Hosts3D) F on OSD for double speed packets (instead of Z)
 - (Hosts3D) Don't create automatic link lines for broadcast packets
 - (Hosts3D) Host Information displays if automatic link lines active
 - (Hosts3D) Fixed possible crash when automatic link lines active
1.11 (29 Jun 10)
 - (Hosts3D) Removed automatic deletion of inactive hosts
 - (Hosts3D) Ctrl + D key now toggles add destination hosts, D on OSD
 - (Hosts3D) Simplified OSD with B indicating simulated broadcasts is active
 - (Hosts3D) No indication that broadcasts are present if simulated broadcasts is off
 - (Hosts3D) Get broadcast addresses from net position entries (255.255.255.255 also)
1.10 (25 Jun 10)
 - (Hosts3D) Check for duplicate packet objects in latest frame
 - (Hosts3D) More efficient check for duplicate active objects in latest frame
 - (Hosts3D) Changed OSD AD to A for active anomaly detection, flashes when anomaly present
 - (Hosts3D) P key now shows packets for selection
 - (Hosts3D) Ctrl + P key now stops showing packets for selection
 - (Hosts3D) Ctrl + T key now toggles show packet destination port
 - (Hosts3D) Ctrl + D key toggles automatic deletion of inactive hosts, X on OSD
 - (Hosts3D) Ctrl + U key toggles show packets for new hosts, N on OSD
 - (Hosts3D) U, P and K no longer displayed on OSD for packet mode
 - (Hosts3D) Host Information displays if anomaly is present, if showing packets and expiry time
 - (Hosts3D) Fixed OSD not updating when protocol or port 0 (all) entered
 - (Hosts3D) Fixed host disappearing (show on-active) when it has an anomaly
 - (hsen)    FreeBSD support
1.09 (06 Jun 10)
 - (Hosts3D) Replaced freeglut (GLUT - Mac OS X) with GLFW
 - (Hosts3D) No longer require pthreads
 - (Hosts3D) Fixed 100% CPU usage
 - (Hosts3D) Fixed make host triggering an anomaly
 - (Hosts3D) Common controls across all OSes
 - (Hosts3D) Fixed crash when closing in Windows
 - (Hosts3D) Mouse scroll wheel works in Mac OS X
 - (Hosts3D) Unable to move hosts with Ctrl + mouse movement
 - (Hosts3D) No cross cursor
 - (Hosts3D) Anomaly alert shape now brackets
 - (Hosts3D) Start/Stop local hsen sudo command not auto-generated
 - (Hosts3D) F1-F4 keys now shows packets from sensors 1-4
 - (Hosts3D) F5 key now shows packets from all sensors
 - (Hosts3D) F7 key now opens a packet traffic file
 - (Hosts3D) F8 key now saves a packet traffic file
1.08 (07 Feb 10)
 - (Hosts3D) Fixed host not disappearing (show on-active) when packet killed before reaching host
 - (Hosts3D) Stopped host disappearing (show on-active) when it is in selection
 - (Hosts3D) Putting 'hold' as colour for net position entry will place hosts in same position
1.07 (30 Nov 09)
 - (Hosts3D) M key displays make host window
 - (Hosts3D) Fixed selection box position with crosshair cursor
 - (Hosts3D) Minimised sleep timers
 - (Hosts3D) Fixed possible services anomaly detection inactivity
 - (Hosts3D) Fixed pause animation disrupting input cursor and record/replay time
 - (Hosts3D) Fixed identifying default 2D GUI button when multiple exist
 - (hsen)    Allow 0.0.0.0 destination address
1.06 (25 Oct 09)
 - (Hosts3D) Controls window now called help window
 - (Hosts3D) Put packet legend in help window
 - (Hosts3D) Improved font readability
 - (Hosts3D) Put "posse" logo in about window
 - (Hosts3D) Fixed possible anomaly detection inactivity
 - (Hosts3D) Changed default settings to show all packets and alert on-active
 - (Hosts3D) Display red AD when anomaly detection active
 - (Hosts3D) Improved 2D GUI scrolling
 - (Hosts3D) Stopped hiding background text when menu displayed
 - (Hosts3D) Fixed compile errors by adding option -lX11
1.05 (06 Oct 09)
 - (Hosts3D) More efficient font and font display system
 - (Hosts3D) Modified 2D GUI to accommodate new font
 - (Hosts3D) Flashing cursor for text input
 - (Hosts3D) Changed mouse pointer to arrow in 2D GUI
 - (Hosts3D) Free OpenGL display lists on exit
 - (Hosts3D) Replaced red anomaly detection active box with red letter A
1.04 (21 Sep 09)
 - (All)     Mac OS X support
 - (Hosts3D) Fixed XOpenDisplay by including XCloseDisplay when displaying fullscreen
 - (Hosts3D) Replaced pkill with killall in default stop local hsen command
 - (Hosts3D) Use atexit to run cleanup on close
 - (Hosts3D) Added promiscuous mode checkbox to start local hsen
 - (hsen)    Use total length in IP header for packet size
 - (hsen)    By default disable promiscuous mode, option -p to enable
 - (hsen)    Install in /usr/local/bin/

src/objects.h
/* objects.h - 26 Jul 10
   Hosts3D - 3D Real-Time Network Monitor
   Copyright (c) 2006-2010  Del Castle

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details. */

#include <sys/time.h>  //timeval

#include "common.h"
#include "misc.h"

#define HSD_EDTNAME  0x01  //edit hostname
#define HSD_EDTRMKS  0x02  //edit host remarks
#define HSD_EXPTCSV  0x03  //export CSV
#define HSD_FNDHSTS  0x04  //find hosts
#define HSD_HNLOPEN  0x05  //open network layout file
#define HSD_HNLSAVE  0x06  //save network layout file
#define HSD_HPTOPEN  0x07  //open packet traffic file
#define HSD_HPTSAVE  0x08  //save packet traffic file
#define HSD_HSTINFO  0x09  //host info
#define HSD_MAKEHST  0x0A  //make host
#define HSD_NETPOS   0x0B  //edit netpos
#define HSD_PKTSPRO  0x0C  //packets protocol
#define HSD_PKTSPRT  0x0D  //packets port
#define HSD_SETCMDS  0x0E  //set commands
#define HSD_SLINACT  0x0F  //select inactive
#ifndef __MINGW32__
#define HSD_HSENRUN  0x10  //run hsen
#define HSD_HSENSTP  0x11  //stop hsen
#endif

const int SPC = 32, MOV = 3, HPR = 20, SELBUF = 50000, SVCS = 32;  //host spacing, movement spacing, hosts per row, host selection buffer, services per host
const double DEPTH = 8192.0;  //perseptive depth
const char tipn[3][6] = {"IP", "IP/Nm", "Nm"}, tips[3][4] = {"Off", "Sel", "All"}, tona[5][6] = {"Off", "Alert", "IP/Nm", "Host", "Sel"}
  , tclr[10][8] = {"default", "orange", "yellow", "fluro", "green", "mint", "aqua", "blue", "purple", "violet"};
const GLubyte bitmaps[17][8] =
{
  { 31,  31,  31,  15,  15,   7,   3,   0},  //posse body 1
  {248, 248, 248, 240, 240, 224, 192,   0},  //posse body 2
  { 15,   7,   3,   3,   7,  15,  15,  31},  //posse body 3
  {240, 224, 192, 192, 224, 240, 240, 248},  //posse body 4
  {111, 239, 239, 239, 255, 255, 127, 127},  //posse body 5
  {246, 247, 247, 247, 255, 255, 254, 254},  //posse body 6
  { 47,  47,  47,  47,  47, 111, 111, 111},  //posse body 7
  {244, 244, 244, 244, 244, 246, 246, 246},  //posse body 8
  { 32,  32,  32,  16,  16,   8,   4,   3},  //posse edge 1
  {  4,   4,   4,   8,   8,  16,  32, 192},  //posse edge 2
  {112,   8,   4,   4,   8,  16,  16,  32},  //posse edge 3
  { 14,  16,  32,  32,  16,   8,   8,   4},  //posse edge 4
  { 72, 136, 136, 136, 128, 128,  64,  64},  //posse edge 5
  { 18,  17,  17,  17,   1,   1,   2,   2},  //posse edge 6
  { 80,  80,  80,  80,  80, 144, 144, 144},  //posse edge 7
  { 10,  10,  10,  10,  10,   9,   9,   9},  //posse edge 8
  {254, 254, 254, 254, 254, 254, 254,   0}   //box
};

//GL position
struct pos_type
{
  double x, y, z;
};

//GL vertex
struct vtx_type
{
  float x, y, z;
};

//GL triangle
struct tri_type
{
  unsigned char a, b, c;
};

//GL square
struct sqr_type
{
  unsigned char a, b, c, d;
};

//GL view direction
struct view_type
{
  double ax, ay;  //angle
  pos_type ee, dr;  //eye, direction
};

//GL cross object
struct cobj_type
{
  vtx_type vtx[28];
  sqr_type sqr[8];
};

//GL host object
struct hobj_type
{
  vtx_type vtx[10];
  tri_type tri[16];
};

//GL multiple-hosts object
struct mobj_type
{
  vtx_type vtx[11];
  tri_type tri[16];
};

//GL packet object
struct pobj_type
{
  vtx_type vtx[8];
  tri_type tri[12];
};

//host object info
struct host_type
{
  unsigned char lsn, clr, vis, sld:1, pip:1, anm:1, shp:1, lck:1, alk:1, spr:2;
    //last sensor, colour, visible, selected, persistant IP, anomaly, show packets, lock, auto link, spare (future use)
  int px, py, pz;
  unsigned long long dld, uld;  //downloads, uploads
  time_t lpk;  //last packet
  host_type *col;
  in_addr hip;
  char htip[16], htmc[18], htnm[256], htrm[256];  //IP, MAC, name, remarks
  int svc[SVCS];  //service
};

//link object info
struct link_type
{
  host_type *sht, *dht;  //source, destination host
  unsigned char spr;  //spare (future use)
};

enum aobj_type { anom, actv, text };  //alert object - anomaly, active, text

//alert object info
struct alrt_type
{
  aobj_type ao;  //alert object
  unsigned char pr;
  unsigned short sz, frm;  //size/duration, frame
  host_type *ht;
};

//packet object info
struct pckt_type
{
  unsigned char pr;
  unsigned short dstpt, frm;  //frame
  pos_type cu;  //current
  host_type *ht;  //destination host
};

enum ptrc_type { stp, hlt, rec, rpy };  //packet traffic - stop, halt, record, replay

//packet traffic record
struct pkrc_type
{
  timeval ptime;  //time of packet
  pkif_type pk;  //packet
};

enum sipn_type { ips, ins, nms };  //display - IP, IP/name, name only
enum sips_type { off, sel, all };  //IP/name display - off, selection, all
enum sona_type { don, alt, ipn, hst, slt };  //on-active - do nothing, alert, IP/name, host, select

//hosts3d settings
struct sett_type
{
  bool hspm;  //start hsen promiscuous mode
  unsigned char sen, pr, mvd;  //display packets from hsen (0 for all), display packets by protocol (0 for all), master visibility duration
  unsigned short prt, osd:1, bct:1, fsp:1, adh:1, anm:1, spe:1, nhl:1, nhp:1, pdp:1, spr:7;  //display packets by port (0 for all)
    //OSD, broadcasts, fast packets, add dest hosts, anomaly detect, spare, new host link, new host packets, packet dest port, spare (future use)
  unsigned int pks;  //number of packets allowed on screen before drop
  char clip[256], hsst[128], hsid[4], hsif[256], hshd[16], hssp[256], cmd[4][256];  //input text clipboard, start/stop hsen, commands
  sipn_type sipn;  //display IP/name
  sips_type sips;  //IP/name display
  sona_type sona;  //on-active
  view_type vws[5];
};

//net position/colour
struct netp_type
{
  int px, py, pz;
  char clr, nt[19];  //colour, CIDR net
};

//draw cross object
void cobjDraw();

//draw host object
void hobjDraw(int r, int g, int b);

//draw multiple-hosts object
void mobjDraw(bool sel);

//draw packet object
void pobjDraw(int c);

//draw alert object
void aobjDraw(bool an);

//return path with hosts3d data directory
char *hsddata(const char *fl);

//add service to host
void svcAdd(host_type *sht, int svc, bool anm);

//load netpos file into memory, create empty
void netpsLoad();

//destroy net positions LL
void netpsDestroy();

//check if an IP address is a broadcast address
in_addr_t isBroadcast(in_addr ip);

//check if a host is to be positioned into a net
int hostNet(host_type *ht);

//copy packet traffic file from/to
void pkrcCopy(const char *ifn, const char *ofn);

//check for controls file, create it
void checkControls();
